### PR TITLE
Checks Filters For Cross-Station Messages

### LINKED ISSUE #63206: Outgoing messages to other stations don't seem to be subject to IC chat filters
Reporting client version: 514.1561

## Round ID:
[174046](https://scrubby.melonmesa.com/round/174046)

## Reproduction:

IC chat filters don't seem to be applied to messages sent to other servers. Seems like an oversight. Probably applies to other types of announcement as well.

I censored the image so I don't get banned by GitHub for the slur but you can probably guess what it is. This was sent on [Terry](https://scrubby.melonmesa.com/round/174046/files?file=game.txt#undefined-3313).

### PR DESCRIPTION
**Pull request by san7890**
## About The Pull Request

Hey there,

I thought this was a bit silly that these messages could bypass the filter, so let's fix that. I used the same principles that we use for AI laws as a reference for what I should be filtering for- let me know if I went outside the use case for these filters or should be more restrictive in filtering.

The gist is, if you add a filtered word to your cross-station message, it'll either be hard-stopped from sending if hard-filtered. If soft-filtered, you'll get a TGUI alert asking you to confirm, and then it'll take a bit longer on the source server and the target server for an admin to review it. This is just simply to ensure that admins can instantly know if it is going to be a detriment to the server, and have more time to get around to it.

Let me know if anything I did in this PR was cringe, I just sorta did what I could to ensure that everything stayed consistent.

## Why It's Good For The Game

Fixes #63206. Feels silly to have such a gaping hole in the filter.

## Changelog

:cl:
admin: Cross-server messages are now filtered against the server's word filter. If it's hard filtered, the message will not send. If it's soft-filtered, it will send slower to allow an admin more time to cancel it.
/:cl:

### CODE REVIEW

**Comment by Mothblocks** [comment]
It doesn't look like you're setting soft_filtering to TRUE anywhere?

**Comment by san7890** [comment]
> It doesn't look like you're setting soft_filtering to TRUE anywhere?

i frogor

**Comment by ATH1909** [comment]
what exactly is this filter checking for? just slurs? netspeak?

**Comment by Mothblocks** [comment]
both

**Comment by optimumtact** [comment]
ratio

**Comment by san7890** [comment]
Filter works, but something's weird with the delay messages stuff that I didn't personally see since I deadminned to bypass the filter and forgot to readmin.

**Comment by san7890** [comment]
image

image

Alright, 75% of it works! I'll take it out of draft when I can verify it's all performant.

**Comment by san7890** [comment]
image

It finally works! Yippie! Wahoo!

**Comment by optimumtact** [comment]
took yo ass 171 rounds for real

**Comment by san7890** [comment]
> took yo ass 171 rounds for real

creature of sloth

File: code/datums/world_topic.dm [hunk 1]
```diff
@@ -102,20 +102,25 @@
...
-	)
+	var/extended_time_display = DisplayTimeText(EXTENDED_CROSS_SECTOR_CANCEL_TIME)
+	var/normal_time_display = DisplayTimeText(CROSS_SECTOR_CANCEL_TIME)
+
+	var/message = "<b color='orange'>CROSS-SECTOR MESSAGE (INCOMING):</b> [input["sender_ckey"]] (from [input["source"]]) is about to send \
+			the following message (will autoapprove in [soft_filter_passed ? "[extended_time_display]" : "[normal_time_display]"]): \
+			<b><a href='?src=[REF(src)];reject_cross_comms_message=[timer_id]'>REJECT</a></b><br><br>\
+			[html_encode(input["message"])]"
+
+	if(soft_filter_passed)
```
Comment (dragomagol) [inline code/datums/world_topic.dm:120]: I was seeing this message even when the cross-server message was "hi"

Comment (san7890) [reply code/datums/world_topic.dm:120]: i hope hi isn't on the soft filter

i'll look into that though, thanks for catching it.

File: code/datums/world_topic.dm [hunk 2]
```diff
@@ -103,7 +103,9 @@
 	// To solve this, we just use a list that we mutate later.
 	var/list/data = list("input" = input)
 	// Did we have to pass the soft filter on our origin server? Passed as a boolean value.
-	var/soft_filter_passed = input["is_filtered"]
+	var/soft_filter_passed = FALSE
+	if(input["is_filtered"])
+		soft_filter_passed = TRUE
```
Comment (Mothblocks) [inline code/datums/world_topic.dm:108]:
```suggestion
	var/soft_filter_passed = !!input["is_filtered"]
```

This makes the intent obvious, protecting this from getting "refactored" later

**Comment by san7890** [comment]
image

Alright, should be good on that end now.

**Pull request merged**